geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderAndCQDurableClientDUnitTest.java

```diff
@@ -0,0 +1,530 @@
+    startDurableClient();
+    createDurableCQs(REGION_CQ, COLO_REGION_CQ);
+
+    verifyGatewaySenderState(true, false);
+
+    // Do some puts and check that data has been enqueued
+    Set<String> keysQueue =
+        clientSite2.invoke(() -> doPutsInRangeTransactionWithTwoPutOperation(0, 30));
+    server1Site2.invoke(() -> checkQueueSize("ln", keysQueue.size() * 2));
+
+    server1Site2.invoke(() -> validateBucketToTempQueueMap("ln", true));
+    server2Site2.invoke(() -> validateBucketToTempQueueMap("ln", true));
+    server3Site2.invoke(() -> validateBucketToTempQueueMap("ln", true));
+
+    // Check that durable client has received all events
+    checkCqEvents(keysQueue.size() * 2);
+  }
+
+  /**
+   * This test case verifies that the server during bucket recovery enqueues all events intended for
+   * that bucket in temporary queue, and that after bucket redundancy is restored events are
+   * transferred from temporary queue to real bucket queue.
+   */
+  @Test
+  public void testSubscriptionQueueWanTrafficWhileRebalanced() throws Exception {
+    configureSites("3", false);
+    verifyGatewaySenderState(true, false);
+
+    List<MemberVM> allMembers = new ArrayList<>();
+    allMembers.add(server1Site2);
+    allMembers.add(server2Site2);
+    allMembers.add(server3Site2);
+
+    // Do some puts so that all bucket are created
+    Set<String> keysQueue = clientSite2.invoke(() -> doPutsInRange(0, 100));
+    server1Site2.invoke(() -> checkQueueSize("ln", keysQueue.size()));
+
+    // check that bucketToTempQueueMap is empty on all members
+    for (MemberVM member : allMembers) {
+      member.invoke(() -> validateBucketToTempQueueMap("ln", true));
+    }
+
+    // Choose server with bucket to be stopped
+    MemberVM serverToStop = getServerToStop(allMembers);
+    int bucketId = serverToStop.invoke(this::getPrimaryBucketList).iterator().next();
+    serverToStop.stop(true);
+
+    // remove from list the member that has been previously stopped
+    allMembers.remove(serverToStop);
+
+    startDurableClient();
+    createDurableCQs("SELECT * FROM " + Region.SEPARATOR + REGION_NAME);
+
+    // configure hook on running members
+    for (MemberVM member : allMembers) {
+      configureHooksOnRunningMember(member, bucketId);
+    }
+
+    // perform rebalance operation to trigger bucket redundancy recovery on running servers
+    String command = new CommandStringBuilder(CliStrings.REBALANCE)
+        .getCommandString();
+    gfsh.executeAndAssertThat(command).statusIsSuccess();
+
+    for (MemberVM member : allMembers) {
+      // All members after redundancy is recovered should have empty temporary queue
+      member.invoke(() -> validateBucketToTempQueueMap("ln", true));
+      // If hook has been triggered on member, then check if member temporarily queued
+      // events while getting initial image from primary server
+      if (member.invoke(ParallelGatewaySenderAndCQDurableClientDUnitTest::isHookTriggered)) {
```

Comment:
Hi @DonalEvans ,

Thanks for the review!

The hook **will be always triggered** on at least on one of the servers that will restore bucket redundancy. Hook is installed on all running servers and then rebalance command is triggered. This code loops through all servers to find the servers that actually restored redundancy (the ones that triggered hook isHookTriggered == true), and then verify that events has been stored to temporary gateway sender queue during recovery of bucket redundancy. So I don't think that this test case could be flaky, since hook will be always triggered on at least one server.

Reply:
Aah, I see, thank you for the explanation.